coordinator/src/orderbook/mod.rs

```diff
@@ -30,120 +41,317 @@ pub mod websocket;
+                    // applied to the orderbook state by joining the trades.
+                    let mut remaining_order = order;
+                    remaining_order.quantity -= quantity;
+                    self.add_order(remaining_order);
 
-pub struct Longs(Vec<Order>);
+                    // update the matched order quantity with the searched for quantity.
+                    order.quantity = quantity;
+                    matched_orders.push(order);
 
-impl Longs {
-    pub fn add(&mut self, order: Order) {
-        self.0.push(order);
-        self.sort();
+                    // we found enough liquidity in the order book to match the order.
+                    break;
+                }
+                Ordering::Equal => {
+                    // we found a perfect match for the searched for quantity.
+                    matched_orders.push(order);
+                    break;
+                }
+            }
+        }
+
+        matched_orders
     }
-    pub fn remove(&mut self, order_id: Uuid) {
-        self.0.retain(|o| o.id != order_id)
+    /// rollback matched orders
+    pub fn rollback_matched_orders(&mut self, matched_orders: Vec<Order>) {
+        for order in matched_orders {
+            if let Some(order) = self.orders.iter_mut().find(|o| o.id == order.id) {
+                order.quantity += order.quantity;
+            } else {
+                self.add_order(order);
+            }
+        }
     }
-    pub fn sort(&mut self) {
-        self.0.sort_by(|a, b| {
+
+    /// Sorts the orders ascending if long and descending if short,
+    fn sort(&mut self, direction: commons::Direction) {
+        self.orders.sort_by(|a, b| {
             if a.price.cmp(&b.price) == Ordering::Equal {
                 return a.timestamp.cmp(&b.timestamp);
             }
-            // Ascending order.
-            a.price.cmp(&b.price)
-        });
+            match direction {
+                // Descending order.
+                commons::Direction::Short => b.price.cmp(&a.price),
+                // Ascending order.
+                commons::Direction::Long => a.price.cmp(&b.price),
+            }
+        })
     }
 }
 
 impl Orderbook {
-    pub fn new(pool: Pool<ConnectionManager<PgConnection>>) -> Self {
+    pub fn new(
+        pool: Pool<ConnectionManager<PgConnection>>,
+        order_matching_fee_rate: Decimal,
+    ) -> Self {
         Self {
             pool,
-            shorts: Shorts(vec![]),
-            longs: Longs(vec![]),
+            shorts: OrderbookSide::new(),
+            longs: OrderbookSide::new(),
+            order_matching_fee_rate,
         }
     }
 
+    /// Initializes the orderbook with non expired open limit orders.
     async fn initialize(&mut self) -> Result<()> {
         let all_orders = spawn_blocking({
             let mut conn = self.pool.clone().get()?;
             move || {
+                // TODO(holzeis): join with trades to get partially matched orders.
```

Comment:
Is this needed? You are updating the order::quantity so that you have already the partially matched  orders.

Reply:
The matched order quantities are only updated in memory but not in the database. When initialising we need to get load how much has already been matched.

Reply:
Got it, thanks.